### PR TITLE
Refactor sphinx.environment.Environment

### PR DESCRIPTION
**Pull request by tk0miya**
To slim down `sphinx.environment.Environment`, I separated indexentries and toctree management feature to some manager classes.
I believe this makes the `Environment` class simple and easy to modify and test them.

### CODE REVIEW

File: sphinx/environment/__init__.py [hunk 1]
```diff
@@ -214,6 +211,14 @@ def __init__(self, srcdir, doctreedir, config):
         # attributes of "any" cross references
         self.ref_context = {}
 
+        self.init_managers()
+
+    def init_managers(self):
```
Comment (tk0miya) [inline sphinx/environment/__init__.py:213]: I could not found good naming for these modules.
I think "manager" is not a best name. So I would be glad if any advice.

File: sphinx/environment/__init__.py [hunk 2]
```diff
@@ -1070,142 +1057,19 @@ def create_title_from(self, docname, document):
...
-            if entries:
-                return nodes.bullet_list('', *entries)
-            return []
-        toc = build_toc(document)
-        if toc:
-            self.tocs[docname] = toc
-        else:
-            self.tocs[docname] = nodes.bullet_list('')
-        self.toc_num_entries[docname] = numentries[0]
+        self.managers['toctree'].note_toctree(docname, toctreenode)
```
Comment (tk0miya) [inline sphinx/environment/__init__.py]: To keep compatibility of Sphinx APIs, I added bridge-methods to manager modules.
But I feel this is very strong coupled. We need to improve this in future.

File: sphinx/environment/__init__.py (hunk 1 above)
Comment (tk0miya) [reply sphinx/environment/__init__.py:213]: It seems these managers should be detached on `topickle()`.

Comment (willingc) [reply sphinx/environment/__init__.py:213]: @tk0miya Perhaps `handler` instead of `manager`

Comment (tk0miya) [reply sphinx/environment/__init__.py:213]: @willingc  Thank you for comment.

But the main concern of the modules is not event handling. The environment instance delegates a part of its job to these modules.

The IndexEntries module does following operations:
- collects the index information from docs
- allows to refer them
- builds index pages for some buildes.

The Toctree module does following:
- collects ToC-tree information from docs
- assigns section and figure numbers using toctree info
- builds toctree for some builders
- builds sidebar for some builders (mainly HTML builders)

Thanks,

**Comment by tk0miya** [comment]
Without naming problem, there are no responses to this PR.
So I will merge this soon.

**Comment by tk0miya** [comment]
Merged. Thank you for reviewing!

**Pull request merged**